LSPS4: Continuous JIT Channels.

ZmnSCPxj-jr opened the pull request:
Alternative to #22 

Closes: #25 

Closes: #26 

Much more complex than the currrent LSPS2 spec.

Note: we really should not be clearing state based on disconnections of the peer, because CLN, LND, and LDK have race conditions between connection/disconnection notification and receiving and sending of custom messages.  This is why this is unnecessarily complex as both sides kinda need to persist information across connection sessions for robustness against this race condition.

Advantages:

* Continuous JIT: keep opening new channels as your inbound liquidity depletes, with the client not having to keep track of this inbound liquidity (which it might not be able to manage allocation, since e.g. asynchronous receives are intended to be spontaneous and not require a single invoice from the receiver).  Client continuously checks the LSP opening feerates (which can change over time depending on mining fee estimates) as needed.
* Privacy with BOLT11: Can issue multiple ephemeral SCIDs, the client can generate multiple "node ID" aliases of itself so each BOLT11 invoice it issues has a different SCID and payee node ID.
* Onion oracle: Support MPP *and* var-amount invoices.  Even better privacy when PTLCs come into the picture, as the LSP cannot even know if you are receiving a single MPP or multiple unrelated single payments, it only knows the total aggregate.
* Aggregate multiple unrelated payments: If you receive a lot of small payments in a short span of time, each of which is unable to pay for a single channel open, you can aggregate them and receive them all at once to pay for the channel open.
  * This would be significantly improved by asynchronous receives, which we already prepare for in the API.

The major disadvantage is a big increase in complexity in both LSP and client sides.  Not even sure this is implementable with current node software LOL.

ZmnSCPxj-jr committed:
LSPS4: Continuous JIT Channels.

JssDWt reviewed (commented):
I think a process like this with the pending payment parts would be great. It allows for several things
- multipart payments
- 0 amount invoices or overpayment in general
- wallets that send probes before the actual payment

All in all this seamlessly integrates with most functionality available in the lighting network today.

Comments below on how to structure the messaging. I think the queue structure could be removed if the client is notified of every ppp.

ZmnSCPxj-jr commented on LSPS2/README.md:
```diff
@@ -0,0 +1,1001 @@
+This is a "split multipath payment" that cannot be normally
+resolved.
+
+To get around this issue, the client uses the `lsps2.holdincoming`
+API call, then fails the 7 parts using `mpp_timeout`, which should
+cause the payer to re-attempt the payment for those parts.
```
> Another option here is to make the `lsps2.holdincoming` flow default for all htlcs

Hmm, you mean even if the client has sufficient inbound liquidity, the LSP will still hold on to HTLCs and put them into PPPs regardless?

ZmnSCPxj-jr replied:
That does kinda simplify things: always just put anything that goes to a Continuous JIT permanent or ephemeral SCID to a PPP.  We can then rename `retryppp` to `forwardppp` instead.

JssDWt replied:
> Hmm, you mean even if the client has sufficient inbound liquidity, the LSP will still hold on to HTLCs and put them into PPPs regardless?

Yeah exactly. Although technically maybe you wouldn't call it 'holding' the htlc. There's just an additional messaging exchange involved in routing them.

ZmnSCPxj-jr replied:
Forwarding an HTLC also involves holding the inbound HTLC anyway, so I do not find an issue with the terminology, but I think I like this overall idea.

ZmnSCPxj-jr replied:
Done.

ZmnSCPxj-jr commented on LSPS2/README.md:
```diff
@@ -0,0 +1,1001 @@
+> LSP that it has received, processed, and persisted a PPP
+> before proceeding with the next one.
+
+### 4. Managing Pending Payment Parts
```
I wanted to avoid replaying on reconnection, considering the possibility that the paye***r*** might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection.  But maybe that is low-probability enough that it will not matter in practice?

ZmnSCPxj-jr replied:
>It would need some handling for htlcs timeouts, especially for multipart payments.

What special handling would this need?  `failppp` would allow clients to do something like send a `mpp_timeout` failure.

JssDWt replied:
> I wanted to avoid replaying everything on reconnection, considering the possibility that the paye_**r**_ might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection. But maybe that is low-probability enough that it will not matter in practice?

JssDWt replied:
> I wanted to avoid replaying everything on reconnection, considering the possibility that the paye_**r**_ might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection. But maybe that is low-probability enough that it will not matter in practice?

I understand. Most ppps will be handled very quickly with a decision from the client. Basically all ppps need to get a response from the client what to do with it. We could add an intermediate decision for multipart payments, like an ACK, that the client has received it, if it doesn't know what to do with it just yet (if the LSP receives the ack that should mean the client persisted it and it doesn't have to replay that part)

JssDWt replied:
> What special handling would this need? `failppp` would allow clients to do something like send a `mpp_timeout` failure.

I can see how a client can request the LSP to forward the ppp, but when requested, the htlc timed out. In a normal lightning flow the client would not learn about this htlc at all if it timed out, but in this flow the client knows about it and it should be removed from the state machine somehow. Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.

ZmnSCPxj-jr replied:
The queue is basically that ACK that the client has received it --- calling `popthenpeekppp` is basically "ACK the current PPP I have persisted it, gimme the next one".  If we want to ACK at all then I think a queue makes better sense.

I like not having a queue and just replaying everything on reconnect; if we replay everything that the client has not handled yet then the client does not have to persist anything at all, everything is just persisted at the LSP and the client can keep its copy of the PPPs in-memory, not in-disk, because if the client crashes, on reconnect the LSP just re-feeds the entire PPP set to the client.  In case it is the network that is the problem (i.e. client did not crash but mobile user entered a tunnel and lost carrier signal) then the re-send is idempotent, as the client should be able to look them up by PPP identifier and see it already has that PPP in-memory and can disregard the extra notification.

ZmnSCPxj-jr replied:
>Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.

Yes, that is indeed what I plan to do.  Basically, instead of `retryppp` we have a `forwardppp` which accepts an array of PPP identifiers.  `retryppp` has no errors, but `forwardppp` has to have those.

LSP atomically:

* Checks that all identifiers are still in the PPP map.
  * If this check fails, error with `unknown_ppp` (2) with a `data.unknowns` field, an array containing the PPP identifiers that are in the params but not in the PPP map anymore --- basically this is the "I already failed them back".
* Checks that it can get capacity on the channel to the client for all the PPPs, and reserves those capacity for the PPPs.
  * If this check fails, error with `insufficient_funds` (3).  The client SHOULD then proceed to `getopeningfee` and `openwithppp`.
* Otherwise, disables the hold counter (so that the incoming HTLCs are now held) and arranges to forward the specified HTLCs to the client on channels with the client, and returns with a success.

So in this flow, the client always attempts `forwardppp` once it has completed a set of multipath PPPs (or once it gets a PPP that is a single-path payment).  If that fails with `unknown_ppp` it reupdates its in-memory map of PPPs to remove those PPPs that have been failed at the LSP.  If that fails with `insufficient_funds` it then proceeds to `getopeningfee` and (if it likes the opening feerate) `openwithppp`.

JssDWt replied:
> I like not having a queue and just replaying everything on reconnect

I agree

JssDWt replied:
> > Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.
> 
> Yes, that is indeed what I plan to do. Basically, instead of `retryppp` we have a `forwardppp` which accepts an array of PPP identifiers. `retryppp` has no errors, but `forwardppp` has to have those.
> 
> LSP atomically:
> 
> * Checks that all identifiers are still in the PPP map.
>   
>   * If this check fails, error with `unknown_ppp` (2) with a `data.unknowns` field, an array containing the PPP identifiers that are in the params but not in the PPP map anymore --- basically this is the "I already failed them back".
> * Checks that it can get capacity on the channel to the client for all the PPPs, and reserves those capacity for the PPPs.
>   
>   * If this check fails, error with `insufficient_funds` (3).  The client SHOULD then proceed to `getopeningfee` and `openwithppp`.
> * Otherwise, disables the hold counter (so that the incoming HTLCs are now held) and arranges to forward the specified HTLCs to the client on channels with the client, and returns with a success.
> 
> So in this flow, the client always attempts `forwardppp` once it has completed a set of multipath PPPs (or once it gets a PPP that is a single-path payment). If that fails with `unknown_ppp` it reupdates its in-memory map of PPPs to remove those PPPs that have been failed at the LSP. If that fails with `insufficient_funds` it then proceeds to `getopeningfee` and (if it likes the opening feerate) `openwithppp`.

This sounds very good to me.

I wonder whether the part `and reserves those capacity for the PPPs` is doable today or whether we'd need to modify lightning implementations for this. 

What do you mean with `disables the hold counter (so that the incoming HTLCs are now held)`? Do you mean it resumes those htlcs now?

ZmnSCPxj-jr replied:
>I wonder whether the part `and reserves those capacity for the PPPs` is doable today or whether we'd need to modify lightning implementations for this.

If the LSP disallows use of the "real" SCID of the channels to the client, and only allows use of the issued Continuous JIT permanent/ephemeral SCIDs, then the LSP code should be able to keep track of how much of the capacity is in use in the channel by the Continuous JIT system.  Then it simply needs a single counter, protected by some kind of atomicity (a single global lock for all Continuous JIT channels with the client) to reserve this capacity, without having to modify the internals of the Lightning implementation.  This is because only the Continuous JIT system is the one that actually forwards to the clients over those channels, so it should be able to accurately account for every millisatoshi of capacity.

This should be doable at least on CLN --- `htlc_received` hook would let you filter out forwards that go to the "real" SCID, and if you detect an issued permanent/ephemeral SCID, you can then keep track of how much capacity you have in your channel with the client that is not yet reserved for HTLCs.  There is the bit of complication in that actual HTLCs have to pay onchain fees (which means that the amount you have to reserve is "HTLC amount + HTLC onchain fee" --- and you might need to override your onchain fee-reporting in the CLN interfface to the blockchain to at least temporarily stop updating while you are calculating this).  You can even ensure atomicity of the amount being reserved (in case of multiple parallel running `forwardppps` from the client) by using the CLN interface to the database, which includes an incrementing version number you can use to ensure that your data (in this context the capacity of the channel(s) with the client) is not stale before you overwrite it with new data.

I suspect that LND would be broadly similar, and it should be easier with LDK.

> What do you mean with `disables the hold counter (so that the incoming HTLCs are now held)`? Do you mean it resumes those htlcs now?

We write elsewhere that LSPs should hold inbound HTLCs to be forwarded to permanent/ephemeral SCIDs, for 60 to 120 seconds at most.  That is the timer that should be disabled at this point.

ZmnSCPxj-jr replied:
Against this, we should note that the LSP might reserve some outbound liquidity on one or more channels to a set of HTLC, but adding multiple HTLCs is not a single atomic operation, especially if multiple channels are involved.  The following is possible:

* Client+LSP have 2 channels with inbound liquidity.
* Client asks two PPPs to be forwarded in a single `forwardppp` call, intended to be atomic.
* LSP schedules one PPP on one channel, the other PPP on the other channel.
* Client signs off (`commitment_signed`+`revoke_and_ack`) on the new PPP on one channel.
* Before the client can sign off on the other channel it gets closed --- maybe client is messing around, or LSP has automated system (separate from the PPP-handling component) that periodically considers channels for closure.  Thus the channel is closed but with the other PPP not instantaited as an HTLC on that channel.

So we specify LSP SHOULD best-effort send all the HTLCs or PTLCs together, but if is unable to (due to the above race condition), it SHOULD return the H/PTLC(s) that did not make it through to the PPP Map, with new PPP identifiers and with the hold timer continued, and with the notification re-sent.  The client MUST check for this condition occurring.

ZmnSCPxj-jr commented on LSPS2/README.md:
```diff
@@ -0,0 +1,1001 @@
+The LSP opens a channel to the client.
+
+TODO: copy from existing LSPS2.
+
```
TODOs:

The client MUST NOT use the `alias` SCID sent in the 0-conf channel open.

The LSP SHOULD fail the incoming HTLC with `unknown_next_peer` if the `alias` SCID is used as next hop.

Intent is that the client should just use this API entirely.

ZmnSCPxj-jr replied:
Done.

ZmnSCPxj-jr commented on LSPS2/README.md:
```diff
@@ -0,0 +1,1001 @@
+# LSPS2 Continuous JIT Channel
```
TODO: rename to LSPS4

ZmnSCPxj-jr replied:
Done.

ZmnSCPxj-jr commented:
Updated: Always PPP-hold all HTLCs, change `retryppp` to `forwardppps` (note plural), replay notifications at reconnection instead of maintaining a queue of PPP identifiers, renumber to LSPS4.

AnthonyRonning reviewed (commented):
Really enjoy the considerations made here for privacy, MPP, eventual async/ptlc, and long-lasting relationships with clients. Seems like it solves several issues I have considered with LSPS2.

Is the justification between the two specs understood to be that LSPS2 is the simple, first pass approach that LSPs will implement first while working to support LSPS4? Or is LSPS4 supposed to replace LSPS2 (or not even merge LSPS2 at all)?

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,932 @@
+PPPs, after which it should just fail the PPP regardles of
+whether the client has gotten around to processing it or not.
+
+* For HTLC-type PPPs, the hold time SHOULD be 60 to 120 seconds.
```
Seems reasonable.

ZmnSCPxj-jr replied:
Done.

TonyGiorgio commented on LSPS4/README.md:
```diff
@@ -0,0 +1,932 @@
+`type` is a string, either `"p"` to request for a permanent SCID,
+or `"e"` to request for an ephemeral SCID.
+
+`expiry_time` is the time to request.
+Expiry times have granularity up to the millisecond.
```
> If a client asks for a permanent SCID and the LSP remembers one, it should respond with the existing permanent SCID and return the expiry_time of the existing permanent SCID.

Then it doesn't really matter what the client wants if the LSP dictates the `expiry_time` in the end. It's also something that is done before the client creates the `expiry_time` on the invoice, so it doesn't change anything, they need to, no matter what, set the `expiry_time` based on the response by the LSP. Asking for a specific `expiry_time` doesn't mean anything.

Further, `expiry_time` can be in the past. What are clients supposed to do then?

ZmnSCPxj-jr replied:
`expiry_time` is respected by the LSP for ephemeral SCIDs; it is only permanent SCIDs where `expiry_time` may be different from the client-specified `expiry_time`.

If the LSP still remembers the permanent SCID, then the client does nothing further --- the point is to add a permanent SCID that the LSP will remember, and it is immaterial if the `expiry_time` is in the past.

An LSP giving a different `expiry_time` for an ephemeral SCID is an error on the LSP side, i.e. the LSP would not be conforming to the standard.

TonyGiorgio commented on LSPS4/README.md:
```diff
@@ -0,0 +1,932 @@
+The `lsps4.forwardppps` call has the following errors
+defined (error `code`s in parenthesis):
+
+* `nonexistent_ppp` (2) - One or more of the listed `ppps` is
+  not in the PPP Map; the expected case is that the hold
+  timer for the PPP has passed and the LSP has failed the
+  underlying PPP.
```
Sorry, somehow missed  that.

ZmnSCPxj-jr replied:
Resolving

TonyGiorgio commented on LSPS4/README.md:
```diff
@@ -0,0 +1,932 @@
+The client can then determine if enough MPP payment parts have reached
+the the client as actual HTLCs or PTLCs, and as LSP-held PPPs, and
+can then attempt to have the LSP forward all those parts to the client.
+This attempt to forward is done via `lsps4.forwardppps`.
```
Okay that makes sense. Understanding that it's a check and that it will proceed further helps.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,932 @@
+
+The client selects a set of PPPs it wants to have forwarded
+to it, and which the client is willing to spend to pay for
+inbound liquidity.
```
Reworded

ZmnSCPxj-jr commented:
Updated: change API version negotiation scheme, start copying the bits from LSPS2, clarify about `hash` for HTLC-type PPPs.

ZmnSCPxj-jr commented:
Updated: explain why all HTLCs need to be put into PPPs even if the client already has inbound capacity

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1107 @@
+simply "jump into" Lightning and receive their first
+ever Bitcoins on Lightning.
+
+JIT Channels allows the channel opening fee to be paid from
```
done

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1107 @@
+the first payment to the user.
+This allows users to "jump into" Lightning easily.
+
+Continuous JIT Channels is a logical extension of the JIT
```
done

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1147 @@
+
+The client can request two types of SCIDs:
+
+* Permanent.
```
The permanent SCID represents **all** channels between the LSP and the client, not just one of them.  If there is no channel yet, the permanent SCID represents the possibility of a future channel between the client and LSP.

The distinction between permanent and ephemeral is not about channel lifetimes, but privacy.  A simple client might not want to have to handle multiple key pairs, one for each invoice it issues, at the cost of privacy.  In that case it just gets a single permanent SCID from the LSP and can store that without having to ask the LSP (for example it allows the client to generate an invoice without being connected to the LSP, e.g. it is a mobile client that wants to issue an invoice that will be paid by a payer at a later time when the client **is** online, or via an asyncpay).

ZmnSCPxj-jr replied:
> If the user wants permanent SCIDs, it should use it's real SCIDs.

If there is no channel yet, there is no real SCID, and remember, the JIT flow is intended to support the case where the client has no channels yet.  That is why the permanent SCID exists.

As I note as well, the permanent SCID outlives every channel the client and the LSP have; as long as at least one channel exists at a time, you can have a hand-over-hand where you open a new channel and close an old one, and the permanent SCID remains valid.

ZmnSCPxj-jr replied:
Removed ephemeral SCIDs.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1147 @@
+
+Ephemeral SCIDs are provided to give privacy to clients.
+For each invoice, the client requests an ephemeral SCID and generates
+a fresh keypair as the node ID; separate invoices thus have a
```
The point of ephemeral SCIDs is that ***if*** the client wants to make multiple ephemeral node IDs, the LSP does not leak that privacy by using a single SCID with the client.  So it is kinda backwards --- if the client wants to implement ephemeral node IDs, it is allowed to get ephemeral SCIDs.  If it used a permanent SCID or one that is attached to an actual on-chain UTXO, then multiple invoices might be to different node IDs but have the same SCID from LSP to client, defeating the point of ephemeral node IDs.

SeverinAlexB replied:
Ephemeral node IDs are a Route Blinding feature.

In Route Blinding, SCIDs are encrypted in the invoice. [Only the next hop SCID](https://github.com/lightning/bolts/blob/master/proposals/route-blinding.md?plain=1#L88) is available to the intermediate node. Therefore the only node that sees the the ephemeral SCID is the LSP itself which has a Map<ephemeral SCID, nodeId> anyway. Hence, I don't see any advantage for privacy here.

ZmnSCPxj-jr replied:
Ephemeral node IDs are ***not*** a route-blinding-only feature.  They can be used, today, with an LSP, if your LSP can issue ephemeral SCIDs like specified here in LSPS4.  That way, a BOLT11 invoice can have an epehemeral node ID ***and*** an ephemeral SCID, and only the LSP knows which client that is --- everyone other than the LSP and client cannot know it.

ZmnSCPxj-jr replied:
Removed ephemeral SCIDs.

ZmnSCPxj-jr commented:
Updated: Move some text blocks around, explain permanent SCIDs a bit better, remove duplicated paragraph.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1107 @@
+  read them and process them.
+* Client attempts to have the LSP directly forward the PPP(s) of a
+  single payment via the `lsps4.forward_ppps` API call.
+  If this succeeds and all the HEDL HTLCs arrive at the client, the
```
done.

SeverinAlexB commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+> Thus, privacy requires that a client both implement ephemeral node
+> IDs, *and* request ephemeral SCIDs from the LSP.
+
+Permanent SCIDs are provided to give a stable path to clients.
+For example, the upcoming BOLT12 offers will need a stable permanent
+SCID to be embedded in a blinded route.
```
This is a flaw of Route Blinding IMO. Route Blinding [encodes an encrypted path into the invoice](https://github.com/lightning/bolts/blob/master/proposals/route-blinding.md?plain=1#L194) based on SCIDs. SCIDs lifespan is potentially very short, especially in case of mobile wallets with a single channel that gets spliced often.

This flaw should be fixed in the Route Blinding spec, not here. It should encode blinded node ids into these paths. Their lifetime is substantially longer.

ZmnSCPxj-jr replied:
Node IDs are large (33 bytes, 32 if you X-only it).  The BOLT spec specifically uses SCIDs because they are smaller (8 bytes).  We count bytes here because we expect multiple hops in the `update_add_htlc` and the onion size is limited to 1366 bytes total (and that already has to include multiple hops). We also need 32 bytes for MAC per hop already, so using SCIDs cuts down per-hop size quite a bit.

(We should note that route blinding is basically just a copy-paste of the BOLT4 onion encoding; the only substantive change is that in route blinding, the latter part of the BOLT4 onion is encoded using an ephemeral key from the receiver while the earlier part is encoded using an ephemeral key from the sender, whereas the BOLT4 of today has it entirely encoded using an ephemeral key from the sender.  If you want to use node IDs instead of SCIDs here, you need to modify the BOLT4 spec as well, not just the route blinding spec )

ZmnSCPxj-jr replied:
FINALLY ephemeral SCIDs and ephemeral node IDs in LSPS4 are for non-route-blinded BOLT11 use.  The ephemeral SCIDs and ephemeral node ID prevent the payer from learning the identity of the payee using the BOLT11 invoice (the payee might leak their privacy another way, but at least it will not happen by the BOLT11 invoice).  The resulting BOLT11 invoice is payable using standard payers today that have not even implemented route blinding yet.

SeverinAlexB replied:
They should have just used the last 8 bytes of the node id to indicate the next hop if they want to save space 🤷‍♂️ Anyway, I see the complexity of such a change which is impracticle.

What I am trying to do here is to reduce the complexity of the spec by challenging its assumptions.
With what you are saying, I think of the following question:
- Why do we add a privacy feature to BOLT11 if BOLT11 will be replaced by BOLT12 in the near future anyway?

Maybe we just need 1 permanent SCID to solve the "SCID are changing all the time" issue. This would also give us the ability to create invoices offline too. No ephemeral SCIDs needed therefore.

ZmnSCPxj-jr replied:
>Why do we add a privacy feature to BOLT11 if BOLT11 will be replaced by BOLT12 in the near future anyway?

As you have pointed out, these things take time to implement and it is likely that a majority of implemented services will service BOLT11 (because it is already widely supported today) and the uptake of BOLT12 is likely to lag for a few more years.

ZmnSCPxj-jr replied:
As I mentioned in the online meeting:

* The LSP-side API that gives the ephemeral SCIDs is going to pretty much call the same SCID generation function, and ephemeral and permanent and "actual" SCIDs are all in the same namespace and you need to ensure they do not conflict anyway.
* Clients do not need to immediately implement the multiple-node-IDs thing.
* Clients that DO implement the multiple-node-IDs thing can issue an invoice that is 100% seamlessly compatible with existing BOLT11 payers, so even if custodial services remain stuck with BOLT11 and do not support BOLT12, they can still pay to privacy-enhanced BOLT11 invoices.
  * Custodial services not implementing this are fine since they are unlikely to need JIT channels at all, they are likely to have pre-arranged inbound liquidity (e.g. LSPS1) and individual users of the custodial service already get some privacy anyway since all invoices they generate point at the custodial service, not their own individual Lightning node.

ZmnSCPxj-jr replied:
Ephemeral SCIDs have been removed.

SeverinAlexB reviewed (commented):
First review pass (will need more passes in the future). In general, I like the approach. It's definitely more complicated than LSPS2. It's important we are aware of how [Route Blinding](https://github.com/lightning/bolts/blob/master/proposals/route-blinding.md) works to find the right balance between privacy and features.

Thx Zmn for this proposal.

ZmnSCPxj-jr commented on LSPS4/README.md:
````diff
@@ -0,0 +1,1788 @@
+}
+```
+
+`lsps4_scid` is the SCID reserved [<LSPS0.scid>][].
````
I think a more straightforward limit would be to recommend that block be at least 8,388,308 (i.e. top bit of block is set).

Also LDK already implemented their own pick-a-random-SCID.  @tnull any thoughts on this?

Nevertheless this is only a recommendation "SHOULD", in principle the LSP can pick any SCID number.

ZmnSCPxj-jr replied:
Done, went with recommending a "set highest bit to 1, randomize other bits" as a SHOULD.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+Otherwise, if a fresh `"p"`ermanent SCID was created, then the
+`expiry_time` MUST be the same as the request parameter.
+
+`ln_base_fee` is the base fee that the client MUST indicate as the
+base fee for the given SCID [<LSPS0.msat>][].
+`ln_prop_fee` is the proportional fee the client MUST indicate as
+the proportional / parts-per-million fee for the given SCID
+[<LSPS0.ppm>][].
```
Same for the lifetime of the SCID seems simplest.

ZmnSCPxj-jr replied:
Done.  Went with "same for the lifetime" for ephemeral SCIDs, but added a notification to change the settings for permanent SCIDs.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+
+Each PPP has a type.
+Currently, the only defined PPP type is HTLC-type PPP, also
+known as `"type": "h"` PPPs.
```
Sure, could do that.

ZmnSCPxj-jr replied:
Done.  Also changed for `"permanent"` and `"ephemeral"` SCIDs in `reserve_scid`.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+
+### 7. Opening Channel And Forwarding Selected Pending Payment Parts
+
+The LSP requests a channel open to the client via standard
```
Definitely defer splicing to version 2.

I should note that 0-conf splicing is a significant increase in risk among all participants.  Not only can an honest LSP bug cause the splice to be double-spent, an honest client bug or client operator error can cause the splice to be double-spent too (basically if the client broadcasts a pre-splice commitment transaction, then that double-spends an input of the splice transaction, and if that is what miners pick up, the splice never goes through).

In particular, splicing is fraught with peril when you batch.  If you batch, any one of your splicing counterparties can double-spend an input of the splice transaction (using an old pre-splice commitment transaction, and Satoshi help you if you allowed 0 reserves on them and they had no funds to be used for punishment at the time), which cancels ***all*** splices in the same batch.  If you had another operation in the same batched transaction --- e.g. you spliced in with client A who has a channel with you, and 0-conf-open a new channel with client B because client B had no channels yet --- then splice counterparties can cancel all the operations in the same batch.  You cannot batch ***ANY*** operation with splicing, whether or not another splice operation, without this risk.

Single-sided channel opens (openv1) have no such risk, the LSP controls the funding transaction completely, so any issues with 0-conf channel opening, batched or not, are entirely the LSP fault.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+The LSP requests a channel open to the client via standard
+[BOLT2 Channel Establishment][] flow.
+
+The LSP selects the channel size.
```
Yes, that is why the LSP has this decision and why `opening_fee` is based on payment size.  This was already decided on in LSPS2.

SeverinAlexB replied:
We decided how we calculate the fee. We never talked about who determines the channel size as far as I know.

ZmnSCPxj-jr replied:
https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/22#discussion_r1163172128

ZmnSCPxj-jr replied:
Basically:

* Channel size is a cost on the LSP: the larger the channel size, the greater the cost on the LSP (needs more UTXOs, risks locking up more of their liquidity to a user that ends up not using it, etc).
* LSP thus wants cost of JIT channel to be proportional to the capacity of the channel it will make.
* UX is harder if the client software decides channel size by itself without human input.
* Human only cares about payment size, for UX we want client software to hide the channel size decisionmaking from the human.
* Payment size is approximately an estimate on how much capacity the human will need in the future; it is likely that the human will receive more inflow of similar size in the future.
* Thus: human indicates payment size to client software, LSP charges based on payment size (good for UX), LSP now has cost basis for how large a channel it can build to the client.

SeverinAlexB replied:
Cool. Let's try this then. In case there is demand/good reason for client determined channel size, we can always ship a v2.

ZmnSCPxj-jr commented on LSPS4/README.md:
````diff
@@ -0,0 +1,1788 @@
+
+```JSON
+{
+  "ppp": "id#0123456789abcdef",
````
I should note that the format of the PPP identifier is up to the LSP and it is valid for the LSP to use `id#` prefixes, or even suffixes if that is what the LSP wants.

But sure `ppp_id` is more accurate.

SeverinAlexB replied:
Fair. Just confused :)

ZmnSCPxj-jr replied:
Done.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1788 @@
+The payer will now send out HTLC(s) going towards the LSP, with
+the next hop being some LSPS4-reserved SCID.
+
+#### Pending Payment Parts (PPPs)
```
Yeah I am Zman sooo obviously cool names are my specialty.

ZmnSCPxj-jr commented:
Updated: use names instead of one-character codes for types (e.g. `"htlc"` instead of `"h"` ), rename `ppp` / `ppps` to `ppp_id` / `ppp_ids` , recommend how to make the SCIDs, allow LSP to change feerate and CLTV-delta of permanent SCIDs.

ZmnSCPxj-jr commented:
Updated: Added `via_scid` field to PPP notifications and a corresponding  `via_scid` TLV to `update_add_htlc` when forwarding those PPPs, for use with private stateless BOLT11 invoices. @tnull

ZmnSCPxj-jr commented:
Updated: Remove `ephemeral` SCIDs.

SeverinAlexB commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1889 @@
+{
+  "ppp_id": "id#abcdefghijklmnop",
+  "details": {
+    "type": "htlc",
```
Type is redundant here because the LSP already knows the type through the `ppp_id`.

ZmnSCPxj-jr replied:
Better to be explicit than implicit.

SeverinAlexB replied:
Don't see any reason for that. Looks to me like a "duplicated code" code smell.

ZmnSCPxj-jr replied:
Done.

SeverinAlexB reviewed (commented):
All in all, the spec is already pretty solid IMO.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1889 @@
+> them likely to become real SCIDs with peers of the LSP node, even
+> if the counter is started from 0.
+
+`expiry_time` is the expiry time for the reserved SCID
+[<LSPS0.datetime>][].
```
We do still need an expiry, for example if the client is trying to DoS and waste the storage of the LSP by reserving and then disconnecting forever.  The LSP and the client need to agree on how long that period would be before the LSP can safely forget the client if the client never comes back again.

SeverinAlexB replied:
Why do we need to expire the permanent SCID? Can't we just use the same SCID for a node pubkey all the time?
In this case, I don't see how the client can DDoS the LSP as there is the same SCID returned every time.

ZmnSCPxj-jr replied:
>In this case, I don't see how the client can DDoS the LSP as there is the same SCID returned every time.

A client can always generate a new node pubkey, you can literally treat the private key as a large 256-bit number and increment it each time, no need to even use a CSPRNG if you are just using throwaway keypairs for the purpose of DoS.  With e.g. Tor or a botnet the IP address is not the same too.

SeverinAlexB replied:
Not sure I can follow. Let's discuss it on the call today.

ZmnSCPxj-jr replied:
Not resolved in previous call.

@SeverinAlexB proposes that an actual LSP implementation can use a 63-bit hash of the client node ID, so that the permanent SCID does not need to be remembered.

However, some software such as LDK has its own way of doing SCID interception.  LDK gets "first dibs" at the data so LSPs built on top of LDK cannot change the way that LDK creates interception SCIDs.

In addition, something like a hash is a one-way function.  So if the LSP receives an interception SCID, it has to iterate through all connected peers and hash each of their node IDs to get the SCID, then find the correct SCID.  The LSP can cache this.  However, how about disconnected peers?  Ideally if the LSP receives an SCID for a peer that is disconnected, it would wake it up (e.g. via LSPS5).  So LSP needs to remember the SCID->client mapping, and I think we need to nail down how long the mapping is retained by the LSP after the peer disconnects without making a channel.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1889 @@
+> payment and when the client receives it, then the timeout may
+> now be below the normal `min_final_cltv_expiry_delta` of the
+> client.
+> BOLT requires that the payee fail this with
+> `incorrect_or_unknown_payment_details` with a `height` indicating
+> the new current blockheight, but the LSP cannot differentiate
+> between this failure (which would cause the payer to retry with
+> renewed timeouts), versus a deliberate attempt by the payer
+> and client to waste LSP resources (which should cause the LSP to
+> deny future service to this client).
+> By adding a small +2 margin here, we greatly reduce the incidence
+> of such an accident.
```
Hmmm yes I think that is correct, this seems a holdover from LSPS2 and should be removed.

ZmnSCPxj-jr replied:
Ah right, now I remember, there is still a race condition:

* Client checks the incoming CLTV-delta is OK right now.
* Client has no inbound capacity, so does `lsps4.open_with_ppps`.
* Mining network finds a new block!
* LSP opens the channel and sends the HTLCs through.
* The HTLCs now have a lower CLTV-delta than the time the client checked it.

Basically this fixes a time-of-check vs time-of-use race condition.

ZmnSCPxj-jr replied:
Reworded.

ZmnSCPxj-jr commented:
Updated: Remove recommendation on SCID generation, re-add ephemeral SCIDs as an extension.

ZmnSCPxj-jr commented:
Updated: Remove `details.type` from `lsps4.fail_ppp`, revise rationale for the +2 margin in `min_final_cltv_delta_expiry`.

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+normally due to a race condition that changes the liquidity
+towards the client:
+
+* MUST do one of the following:
+  * (preferred) add the PPP back to the PPP Map under a
+    new PPP identifier and resumes its hold timer.
+    It re-sends `lsps4.you_have_new_ppp` under the new PPP
+    identifier.
+    If its hold timer has already expired at this point, the
+    LSP MUST NOT do this, but MUST do the alternative below
+    instead.
+  * (unpreferred) NOT add the PPP back to the PPP Map and
+    fail the payment with `temporary_channel_failure`.
```
These things are hard to accomplish if you're building an LSP on top of CLN or LND. You would forward the htlc through the htlc interceptor interface, and you don't get a second chance. If the channel is not there, it results in unknown_next_peer. Not much you can do about it. (I see this is mostly addressed in the rationale section below)

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+
+The LSP, atomically:
+
+* MUST check that there is currently no pending LSPS4-bought
+  channel currently being opened (i.e. has not completed up to
+  `channel_ready` from both parties).
+  * This is done by checking for the existence of a
+    "PPP Opening Set", described later.
```
Is this strictly necessary?
The main business model of an LSP is to open channels for fees.
As long as this request is not using the same payment parts as a previous request, this should be fine?

ZmnSCPxj-jr replied:
LSP can more easily limit its risk by always handling just one channel open at a time, as well as simplify implementation.  If client cannot rely on an LSP being able to handle multiple channel opens, then it is better if we say that here.

There is a reason why we have a `operation_pending` error code for this.

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+* `insufficient_funds` (5) - The total `payment_size` is not
+  enough to pay for the `opening_fee`.
+
+> **Non-normative** The intent is that once asynchronous receives
+> are possible, asyncpay-type PPPs will NOT be accepted in the
+> `lsps4.open_with_ppps` call.
+> Instead, those PPPs can only be passed to `lsps4.forward_ppps`
+> call, which will cause the LSP to respond to the senders of those
+> asynchronous receives.
+> The senders would then deliver actual HTLCs/PTLCs that the LSP
+> will convert again to PPPs.
```
Interesting!

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+Rejection is via the `error` Lightning BOLT message.
+
+* If the client disconnects before the LSP can receive
+  `fuding_signed` from the client, OR if the client sends an
```
```suggestion
  `funding_signed` from the client, OR if the client sends an
```

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+
+### 8.  Post-opening Normal Operation
+
+The client MUST NOT use the `alias` SCID sent in the 0-conf
+channel open (unless it is equal to the LSPS4 SCID, which
+is **not** assured), or the "real" SCID of the channel once
+confirmed.
+The client MUST use only the SCID issued via `lsps4.reserve_scid`
+for all invoices and offers it creates.
+
+The LSP SHOULD fail an incoming HTLC if it uses the `alias` or
+"real" SCID of the channel as the next hop, with the failure
+`unknown_next_peer`, if the `alias` is not the same as the
+LSPS4 SCID.
+
+> **Rationale** Some node software implementations do not have
+> an API to set the `alias` of the channel.
+> The LSPS4 SCID, in any case, represents all the channels
+> with the client, not any specific channel.
+> It also represents a possible future channel if there is no
+> actual channel between the LSP and the client yet.
```
It could be useful for the client to have the option to bypass the LSPS4 spec by issuing invoices with the real alias scid. In that case it is just a normal forward, without the PPP overhead.

ZmnSCPxj-jr replied:
Client can be simpler if it always just uses the LSPS4 SCID.  It can issue invoices with 0 regard for current inbound liquidity, or if it issues multiple invoices (some of which may or may not be paid, e.g. merchant application) and does not want to have to allocate inbound liquidity to invoices.

For example, suppose the client currently has 1 unit of inbound.  Then it issues two invoices of 1 unit each, with no assurance of whether or not it is going to get paid.  If it used the "real" SCID then if both invoices are paid, then the second one will fail due to lack of capacity.

An LSP can ameliorate this by putting the second one in a PPP even though it did not use the LSPS4 SCID.  This is equivalent to saying that "the real and/or alias SCID of a channel with a client is an alias of the LSPS4 SCID".  But it is much much simpler to just say "client always uses the LSPS4 SCID" for implementation.

JssDWt replied:
Say the client selects a new LSP to get its future channels from, it doesn't want new channels from the old LSP. It can still use the channel with the old LSP, by including a 'normal' hop hint in the invoice, and perhaps another hop hint for the new LSP.

ZmnSCPxj-jr replied:
>Say the client selects a new LSP to get its future channels from, it doesn't want new channels from the old LSP. It can still use the channel with the old LSP, by including a 'normal' hop hint in the invoice, and perhaps another hop hint for the new LSP.

And the "normal" routehint in this case would be the result from `lsps4.get_scid`.

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+The LSP MUST ensure that all parts forwarded sum up to at least
+`payment_size_msat - opening_fee`.
+
+The LSP MUST include an `extra_fee` (type 65536) TLV to the
```
Next to the `extra_fee` TLV, every lsps4 `update_add_htlc` probably needs a `ppp_id` TLV, to correlate the actual HTLC with the PPP.

ZmnSCPxj-jr replied:
Why though?  The PPP exists only as a way to implement onion oracles (where the oracle is the client). In the end an HTLC is an HTLC.

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+  next hop.
+  When payment arrives at the LSP, it creates pending payment parts
+  (PPPs, a fancy invented term for HODL HTLCs) for it.
+* LSP notifies the client with `lsps4.you_have_new_ppp`, indicating that
```
(boring) name suggestion:`lsps4.new_ppp`

ZmnSCPxj-jr replied:
But that is so boring.... "You have new mail!!!"

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+be changed by the LSP at any time, with a grace period during which
+the LSP MUST accept both the previous and current settings.
+The client is notified of this change via the
+`lsps4.i_changed_fee_rates` notification, to be
```
name suggestion: `lsps4.fee_update`

JssDWt commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+`lsps4.reserve_scid` has the following defined errors (error `code`
+in parenthesis):
+
+* `expiry_too_long` (2) - The `expiry_time` requested by the client
```
```suggestion
* `expiry_too_far` (2) - The `expiry_time` requested by the client
```

JssDWt reviewed (commented):
Great work on this proposal!
I've left comments here and there, but the general flow of events is exactly what I would expect from a system like this. Also, the considerations are very thorough.

I like the way it is set up for future protocol improvements, like ptlcs. Also, I was blown away by the (future) asyncpay flow and the option to batch open channels.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+* The LSP node ID.
+* The LSP-provided `cltv_expiry_delta`.
+* The expiry for the payment.
+* Whether it can use MPP or not.
```
Why remove this?  Payer can only use MPP if payee (client in this context) supports it.  In theory a payee can be simplified to the point that it does not support MPP at all, so it should signal whether it supports it or not.  LSPs are forced to always consider the possibility of MPP, as this lets us also consider the case of multiple unrelated tiny payments by using all of them to fund a single channel open.

JssDWt replied:
Ah, I thought this was a remnant of the LSPS2 spec. If the client doesn't support MPP, he can totally put that in the invoice, but I didn't think it was relevant for this spec, because it works regardless of MPP signaling or not.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+* MAY have any format, provided:
+  * MUST be encodable in an ASCII encoded JSON string without any
+    string escapes needed.
+  * MUST not be longer than 24 bytes (not including the `"`
```
Anything finite is fine. Could go 64.  Characters is fine too, as we specify ASCII and in ASCII 1 char = 1 byte.

JssDWt replied:
Let's make it 64 then, to have the hash option.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+then restarts quickly, the client can reconnect and re-download the PPP Map
+from the LSP, improving payment experience.
+
+However, the LSP MAY choose not to persist the PPP Map.
+In this case, if the LSP crashes and restarts, it MUST fail all HTLCs that
+were held in PPPs with `temporary_channel_failure`.
```
Assuming the LSP is able to persist the PPP Map somehow.

In LDK, on restart any intercepted HTLCs are failed automatically with no way for the LSP code to re-send to the client.  So we need to specify this (LDK-based LSPs cannot actually store the PPP Map).

(This is the main reason why we talk about the PPP Map, some software can recover on restart, some (LDK) cannot)

JssDWt replied:
In LND the htlcs are replayed when reconnecting to the grpc stream.
In lspd (breez lsp implementation), we replay htlcs from CLN as well when lspd reconnects.

This allows for a 'stateless' handling of these htlcs. When the node restarts, lspd still has the same state, and ignores replayed htlcs. When lspd restarts, the htlcs are replayed and lspd can continue without remembering the previous state. 

So maybe just change MUST to MAY here, in case the LSP can recover after crash.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+  its PPP type.
+  * Does not respond to the inbound until after the hold timer expires,
+    or the client has handled the payment fully.
+* Adds the PPP to a called the "PPP Map", which maps the PPP identifier
+  to the information the LSP needs to forward, handle, or fail the
+  underlying payment, as well as the information that the client needs to
+  process for that PPP.
+  * If the hold timer expires before the client has decided to forward,
+    fail, or use the PPP for opening a new channel, should also remove
+    it from the PPP Map.
```
"Map" is intended to be a very generic term, not a specific structure.  It is just some mapping from PPP identifier to whatever internal details structure your node software needs.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+In this case, if the LSP crashes and restarts, it MUST fail all HTLCs that
+were held in PPPs with `temporary_channel_failure`.
+
+If the client has previously called `lsps4.reserve_scid` (even on a
+previous connection), this also enables the LSP sending the notification
+`lsps4.you_have_new_ppp`.
```
LSPS0 requires that other LSPS docs specify that notifications are enabled by some client API call.  Thus, we explicitly say that the notification is enabled by calling `reserve_scid`.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+    "amount_msat": "21000",
+    "hash": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
+    "cltv": 599999
+    "onion": "abcdefghijklmnopqrstuvwxyz+/"
```
The difference here is that LSPS4 is a single hop only, between LSP and client.  This is more important for routing nodes, less important for the last hop.

SeverinAlexB replied:
1400 bytes or 2800 bytes. We are still talking bytes *shrug*.

ZmnSCPxj-jr commented on LSPS4/README.md:
```diff
@@ -0,0 +1,1923 @@
+    "hash": "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
+    "cltv": 599999
+    "onion": "abcdefghijklmnopqrstuvwxyz+/"
+  }
```
The expectation is that client is the last hop / ultimate payee; indeed, 0-conf channels (which are what JIT channels build) cannot be published channels.  Thus, the expectation is that the client is always the last hop, and thus there is no way to be the hop that serves as the blinding introduction point.

ZmnSCPxj-jr commented:
TODO: In 2023-11-09 meeting, it was agreed that LSPS4 will specifically state that the SCID returned by `lsps4.reserve_scid` would be different from any SCIDs issued in LSPS2, but without specifying any specific bit-patterns or versioning bits in the SCID.

tnull commented:
Closing this as this repo will be archived shortly. Please reopen/resume work on the bLIP repository.
